### PR TITLE
Support ZBVZeroBubbleSchedule

### PR DESCRIPTION
**Pull request by H-Huang**
This is dependent on the changes in this pytorch stack: https://github.com/pytorch/pytorch/pull/146217

Add support for running `ZBVZeroBubbleSchedule` and v-shaped CSV schedules in torchtitan

Fixes https://github.com/pytorch/torchtitan/issues/774

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -285,22 +290,19 @@ def loss_fn(pred, labels):
...
                 # Pipeline Parallel forward / backward inside step() call
-                is_last_stage = pp_mesh.get_local_rank() == pp_mesh.size() - 1
-
                 with train_context(optional_context_parallel_ctx):
-                    if pp_mesh.get_local_rank() == 0:
-                        pp_schedule.step(input_ids)
-                    elif is_last_stage:
-                        losses = []
-                        pp_schedule.step(target=labels, losses=losses)
+                    targets = labels if has_last_stage else None
```
Comment (wconstab) [inline train.py]: kinda nit picking but i feel like if the stage object inside the schedule already knows that it is first or last, we can avoid having the logic in the training loop too.  

otoh it seems nice to be explicit at the train.py layer on whether we are asking to compute loss or not. 

thoughts? 
@tianyu-l

Comment (tianyu-l) [reply train.py]: It feels nice when we only explicitly pass in meaningful targets/losses when we are not sure if they'll be properly accessed, so I'm OK with these if-else statements.

But how different is `input_ids`? Can we just unify everything into ```pp_schedule.step(input_ids, target=targets, losses=losses)```
and pass `input_ids = None` when `not has_first_stage`?

Comment (H-Huang) [reply train.py]: We can't do `input_ids=None` right now since we have logic that automatically splits all `*args` into microbatches. For example if the user wants to do `step(tensors, None)` that would be split up into microbatches of `(tensor1, None)`, `(tensor2, None)`, ... We could update the splitting logic but not sure if it is worth it

File: train.py [hunk 2]
```diff
@@ -285,22 +290,19 @@ def loss_fn(pred, labels):
...
+                    losses = [] if has_last_stage else None
+                    if has_first_stage:
+                        pp_schedule.step(input_ids, target=targets, losses=losses)
```
Comment (tianyu-l) [inline train.py:296]: what if a schedule has `has_last_stage = True` and `has_first_stage = False` for the output layer -- will it miss the chance to feed in `losses`?

**Review by tianyu-l — COMMENTED** [review]
GPU CI failed, not sure if it is due to the reason I commented.

File: train.py (hunk 2 above)
Comment (H-Huang) [reply train.py:296]: Oops yeah, that was the issue. Updated it and will let the CI run again

**Review by wconstab — APPROVED** [review]

File: train.py [hunk 3]
```diff
@@ -285,22 +290,19 @@ def loss_fn(pred, labels):
...
-                is_last_stage = pp_mesh.get_local_rank() == pp_mesh.size() - 1
-
                 with train_context(optional_context_parallel_ctx):
-                    if pp_mesh.get_local_rank() == 0:
-                        pp_schedule.step(input_ids)
-                    elif is_last_stage:
-                        losses = []
-                        pp_schedule.step(target=labels, losses=losses)
+                    targets = labels if has_last_stage else None
+                    losses = [] if has_last_stage else None
```
Comment (tianyu-l) [inline train.py]: nit
```suggestion
                    targets, losses = (labels, []) if has_last_stage else (None, None)
```

**Review by tianyu-l — APPROVED** [review]
lgtm, had a nit comment
